app/src/main/java/de/westnordost/streetcomplete/quests/NameAndLocationLabel.kt

```diff
@@ -0,0 +1,116 @@
+package de.westnordost.streetcomplete.quests
+
+import android.content.res.Resources
+import android.text.Html
+import android.text.Spanned
+import androidx.core.os.ConfigurationCompat
+import androidx.core.text.parseAsHtml
+import de.westnordost.osmfeatures.FeatureDictionary
+import de.westnordost.streetcomplete.R
+import de.westnordost.streetcomplete.ktx.toList
+import java.util.Locale
+
+fun Resources.getNameAndLocationLabelString(tags: Map<String, String>, featureDictionary: FeatureDictionary): Spanned? {
+    val localeList = ConfigurationCompat.getLocales(configuration).toList()
+    val featureName = getFeatureName(tags, featureDictionary, localeList)
+        ?.withNonBreakingSpaces()?.inItalics()
+    val nameLabel = getNameLabel(tags)?.withNonBreakingSpaces()?.inBold()
+    val levelLabel = getLevelLabel(tags)
+    val houseNumberLabel = getHouseNumberLabel(tags)
+
+    val result = mutableListOf<String>()
+    if (levelLabel != null) {
+        result += levelLabel
+    }
+    // only show housenumber if there is neither name nor level information
+    if (levelLabel == null && nameLabel == null && houseNumberLabel != null) {
+        result += houseNumberLabel
+    }
+    // if both name and feature name are available, put the feature name in parentheses
+    if (nameLabel != null) {
+        result += nameLabel
+        if (featureName != null) result += "($featureName)"
+    } else {
+        if (featureName != null) result += "$featureName"
+    }
+
+    return if (result.isEmpty()) null else result.joinToString(" ").parseAsHtml()
+}
+
+private fun getFeatureName(
+    tags: Map<String, String>,
+    featureDictionary: FeatureDictionary,
+    localeList: List<Locale>
+): String? {
+    val locales = localeList.toMutableList()
+    /* add fallback to English if (some) English is not part of the locale list already as the
+       fallback for text is also always English in this app (strings.xml) independent of, or rather
+       additionally to what is in the user's LocaleList. */
+    if (locales.none { it.language == Locale.ENGLISH.language }) {
+        locales.add(Locale.ENGLISH)
+    }
+    return featureDictionary
+        .byTags(tags)
+        // not for geometry because at this point we cannot tell apart points and vertices
+        // .forGeometry(element?.geometryType)
+        .isSuggestion(false)
+        .forLocale(*locales.toTypedArray())
+        .find()
+        .firstOrNull()
+        ?.name
+}
+
+fun getNameLabel(tags: Map<String, String>): String? {
+    val name = tags["name"]
+    val brand = tags["brand"]
+    val ref = tags["ref"]
+    val operator = tags["operator"]
+
+    return name
+        ?: brand
+        ?: if (ref != null && operator != null) "$operator $ref" else null
+        ?: operator
+        ?: ref
```

Comment:
Do you think it's worth to check whether it's a reference number (e.g. `if (tags["emergency"] == fire_hydrant)`) and show "Reference number %s" in this case?

e.g. https://www.openstreetmap.org/node/6272420812:
![grafik](https://user-images.githubusercontent.com/202916/158060434-aa1ab0e3-3b08-4c3a-bce2-f42c1e8c78d2.png)

But this would likely lead to questions about more and more elaborate logic for this.

Reply:
Hmm, well I suppose there is the number "1" engraved on that fire hydrant right? So if someone looks at the fire hydrant, he will see that the "1" is correct

Reply:
I acknowledge that it looks odd, but as you say, I wouldn't want to introduce special treatments for certain element types